nselib/nbd.lua

```diff
@@ -0,0 +1,583 @@
+    LIST             = 0x00000003,
+    PEEK_EXPORT      = 0x00000004, -- PEEK_EXPORT Extension
+    STARTTLS         = 0x00000005,
+    INFO             = 0x00000006, -- INFO Extension
+    GO               = 0x00000007, -- INFO Extension
+    STRUCTURED_REPLY = 0x00000008, -- STRUCTURED_REPLY Extension
+    BLOCK_SIZE       = 0x00000009, -- INFO Extension
+  },
+
+  opt_rep_types = {
+    ACK                 = 0x00000001,
+    SERVER              = 0x00000002,
+    INFO                = 0x00000003, -- INFO Extension
+    ERR_UNSUP           = 0x80000001,
+    ERR_POLICY          = 0x80000002,
+    ERR_INVALID         = 0x80000003,
+    ERR_PLATFORM        = 0x80000004,
+    ERR_TLS_REQD        = 0x80000005,
+    ERR_UNKNOWN         = 0x80000006, -- INFO Extension
+    ERR_SHUTDOWN        = 0x80000007,
+    ERR_BLOCK_SIZE_REQD = 0x80000008, -- INFO Extension
+  },
+
+  opt_rep_ext_types = {
+    info = {
+      EXPORT   	  = 0x0000,
+      NAME     	  = 0x0001,
+      DESCRIPTION = 0x0002,
+      BLOCK_SIZE  = 0x0003,
+    },
+  },
+
+  cmd_req_flags = {
+    FUA     = 0x0001,
+    NO_HOLE = 0x0002, -- WRITE_ZEROES Extension
+    DF      = 0x0004, -- STRUCTURED_REPLY Extension
+  },
+
+  cmd_req_types = {
+    READ         = 0x0000,
+    WRITE        = 0x0001,
+    DISC         = 0x0002,
+    FLUSH        = 0x0003,
+    TRIM         = 0x0004,
+    CACHE        = 0x0005, -- XNBD custom request types
+    WRITE_ZEROES = 0x0006, -- WRITE_ZEROES Extension
+  },
+
+  errors = {
+    EPERM     = 0x00000001,
+    EIO       = 0x00000005,
+    ENOMEM    = 0x0000000C,
+    EINVAL    = 0x00000016,
+    ENOSPC    = 0x0000001C,
+    EOVERFLOW = 0x0000004B,
+    ESHUTDOWN = 0x0000006C,
+  },
+}
+
+Comm = {
+  --- Creates a new client instance.
+  --
+  -- @name Comm.new
+  --
+  -- @param host Table as received by the action method.
+  -- @param port Table as received by the action method.
+  --
+  -- @return o Instance of Client.
+  new = function(self, host, port)
+    local o = {host = host, port = port, exports = {}}
+    setmetatable(o, self)
+    self.__index = self
+    return o
+  end,
+
+  --- Connects to the NBD server.
+  --
+  -- @name Comm.connect
+  --
+  -- @return status true on success, false on failure.
+  connect = function(self)
+    -- NBD servers send a response when we connect. We are using
+    -- tryssl here as a precaution since there are several
+    -- implementations of the protocol and no reason it can't be
+    -- wrapped. IANA has rejected assigning another port for NBD over
+    -- TLS.
+    local sd, err, proto, rep = comm.tryssl(self.host, self.port, "", {recv_before = true})
+    if not sd then
+      return false, err
+    end
+
+    -- The socket connected successfully over whichever protocol.
+    -- Store the connection information.
+    self.socket = sd
+    self.protocol = {ssl_tls = (proto == "ssl")}
+
+    if #rep ~= 8 then
+      stdnse.debug1("Failed to receive first 64 bits of magic from server: %s", rep)
```

Comment:
Your analysis was correct, the issue was getting sent extra data. I instituted a receive buffer in the protocol object to handle varying sizes of initial receives.